### PR TITLE
Fixed retainability part of Wagner's error recovery

### PR DESCRIPTION
**Pull request by ptersilie**
This PR finally fixes the retainability part of Wagner's error recovery. All standard tests pass and fuzzy tests have been run (and passed) a couple of times as well, suggesting this might be working now. Since this is easy to deactivate again I suggest merging this and see if any bugs manifest in the future.

### CODE REVIEW

**Comment by ptersilie** [comment]
Should be ready for review now.

File: lib/eco/incparser/astree.py [hunk 1]
```diff
@@ -623,10 +624,9 @@ def has_changes(self, version = None):
             return self.get_attr("changed", version) or self.get_attr("nested_changes", version)
         return self.changed or self.nested_changes
 
-    def exists(self, version = None):
-        if version:
-            return not self.get_attr("deleted", version)
-        return not self.deleted
+    def does_exist(self, version = None):
```
Comment (ltratt) [inline lib/eco/incparser/astree.py]: Is it deliberate that the `version` parameter isn't used?

File: lib/eco/incparser/incparser.py [hunk 2]
```diff
@@ -392,63 +397,87 @@ def refine(self, node, offset, error_offset):
...
+        self.pass2(node, offset, error_offset, retain_set)
 
-    def pass2(self, node, offset, error_offset):
+    def pass1 (self, node, offset, error_offset, retain_set):
+        if offset > error_offset:
+            # We don't have to check any other children
+            # that come after the error node
+            return
+        for child in node.get_attr("children", self.prev_version):
+            if (offset + child.textlength() <= error_offset):
```
Comment (ltratt) [inline lib/eco/incparser/incparser.py]: Unnecessary brackets.

**Review by ltratt — APPROVED** [review]

File: lib/eco/incparser/astree.py (hunk 1 above)
Comment (ptersilie) [reply lib/eco/incparser/astree.py]: No, forgot to remove this when I changed the method.

**Comment by ptersilie** [comment]
Addressed your comments.

**Comment by ltratt** [comment]
OK, if you squash, I can merge.

**Comment by ptersilie** [comment]
I didn't manage to squash the cleanup commit since most of it should be squashed into the first commit. I think it's fine to just leave it.

**Comment by ltratt** [comment]
OK.

**Pull request merged**